File: src/com/medium/KthSmallestElementInABST_230.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class KthSmallestElementInABST_230 {
    // Recursive version with memorize
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 3 ms, faster than 10.50%
    // Memory Usage: 38.3 MB, less than 95.53%
    private List<Integer> mList = new ArrayList<>();

    public int kthSmallest(TreeNode root, int k) {
        inorder(root);
        return mList.get(k);
    }

    // inorder traversal: LDR
    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        mList.add(root.val);
        inorder(root.right);
    }

    // Recursive version without memorize
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 37.6 MB, less than 99.71%
    private int mAnswer = 0;
    private int mCnt = 0;

    public int kthSmallest2(TreeNode root, int k) {
        inorder(root, k);

        return mAnswer;
    }

    private void inorder(TreeNode root, int k) {
        if (root == null) {
            return;
        }
        inorder(root.left, k);
        mCnt++;
        if (k == mCnt) {
            mAnswer = root.val;
            return;
        }
        inorder(root.right, k);
    }
}

File: Golang_src/medium/kthSmallestElementInABST_230.go
package medium

// Use inorder traversal with memorize:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 16 ms, faster than 8.20%
// Memory Usage: 6.9 MB, less than 5.33%
func kthSmallest(root *TreeNode, k int) int {
	order := inorder(root)

	return order[k-1]
}

func inorder(root *TreeNode) []int {
	if root == nil {
		return nil
	}
	result := []int{}
	result = append(result, inorder(root.Left)...)
	result = append(result, root.Val)
	result = append(result, inorder(root.Right)...)

	return result
}

// Use stack:
// Time Complexity: O(h+k)
// Space Complexity:O(h+k)
// h: tree height, k: k smallest element
// Runtime: 8 ms, faster than 97.33%
// Memory Usage: 6 MB, less than 38.17%
func kthSmallest2(root *TreeNode, k int) int {
	stack := []*TreeNode{}
	cur := root

	for len(stack) > 0 || cur != nil {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		cur = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		k--
		if k == 0 {
			return cur.Val
		}
		cur = cur.Right
	}

	return 0
}
